### PR TITLE
Adding support to register settings dynamically

### PR DESCRIPTION
**Pull request by ryanbogan**
Signed-off-by: Ryan Bogan <rbogan@amazon.com>

### Description
Adding support to register dynamic settings.
For plugins/extensions, they can define their own settings which are propagated across the cluster.
This PR adds support to register Node Scope or Index Scope settings at anytime.
This feature support unlocks a part of extensions hot swap.

Recreate of PR, but in main: https://github.com/opensearch-project/OpenSearch/pull/3753

### Check List
- [X] New functionality includes testing.
  - [X] All tests pass
- [X] New functionality has been documented.
  - [X] New functionality has javadoc added
- [X] Commits are signed per the DCO using --signoff
- [X] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by saratvemulapalli** [comment]
@reta you had suggestions on this change.

File: server/src/main/java/org/opensearch/common/settings/AbstractScopedSettings.java [hunk 1]
```diff
@@ -121,8 +122,13 @@ protected AbstractScopedSettings(
                 keySettings.putIfAbsent(setting.getKey(), setting);
             }
         }
-        this.complexMatchers = Collections.unmodifiableMap(complexMatchers);
-        this.keySettings = Collections.unmodifiableMap(keySettings);
+        if (FeatureFlags.isEnabled(FeatureFlags.EXTENSIONS)) {
```
Comment (reta) [inline server/src/main/java/org/opensearch/common/settings/AbstractScopedSettings.java:125]: I think this feature should not be related to `EXTENSIONS`, right?

File: server/src/main/java/org/opensearch/common/settings/IndexScopedSettings.java [hunk 2]
```diff
@@ -251,6 +251,10 @@ public IndexScopedSettings copy(Settings settings, IndexMetadata metadata) {
         return new IndexScopedSettings(settings, this, metadata);
     }
 
+    public boolean registerSetting(Setting<?> setting) {
```
Comment (reta) [inline server/src/main/java/org/opensearch/common/settings/IndexScopedSettings.java:254]: Not needed, method already public (https://github.com/opensearch-project/OpenSearch/pull/5495/files#diff-edf64b791aed5932da1f53fa264e42e98bd473f585ceed12892b66cda79c3fd7R153)

File: server/src/main/java/org/opensearch/common/settings/ClusterSettings.java [hunk 3]
```diff
@@ -173,6 +173,10 @@ public ClusterSettings(final Settings nodeSettings, final Set<Setting<?>> settin
         addSettingsUpdater(new LoggingSettingUpdater(nodeSettings));
     }
 
+    public boolean registerSetting(Setting<?> setting) {
```
Comment (reta) [inline server/src/main/java/org/opensearch/common/settings/ClusterSettings.java:176]: Same as https://github.com/opensearch-project/OpenSearch/pull/5495/files#r1044701718

File: server/src/main/java/org/opensearch/common/settings/SettingsModule.java [hunk 4]
```diff
@@ -179,6 +179,29 @@ public void configure(Binder binder) {
...
 
+    /**
+     * Dynamically registers a new Setting at Runtime. This method is mostly used by plugins/extensions
+     * to register new settings at runtime. Settings can be of Node Scope or Index Scope.
+     * @param setting which is being registered in the cluster.
+     * @return boolean value is set to true when successfully registered, else returns false
+     */
+    public boolean registerDynamicSetting(Setting<?> setting) {
+        try {
+            registerSetting(setting);
```
Comment (reta) [inline server/src/main/java/org/opensearch/common/settings/SettingsModule.java:190]: This part we have not implemented in a reliable way: if `clusterSettings.registerSetting`  or `indexScopedSettings.registerSetting` fails later on, the `setting` will still be registered by `registerSetting` but not applied properly.

File: server/src/main/java/org/opensearch/common/settings/AbstractScopedSettings.java (hunk 1 above)
Comment (dblock) [reply server/src/main/java/org/opensearch/common/settings/AbstractScopedSettings.java:125]: +1, we can just enable this IMO

File: server/src/main/java/org/opensearch/common/settings/SettingsModule.java [hunk 5]
```diff
@@ -179,6 +179,34 @@ public void configure(Binder binder) {
...
+        boolean onNodeSetting = false;
+        boolean onIndexSetting = false;
+        try {
+            if (setting.hasNodeScope()) {
+                onNodeSetting = clusterSettings.registerSetting(setting);
+            }
+            if (setting.hasIndexScope()) {
+                onIndexSetting = indexScopedSettings.registerSetting(setting);
+            }
+            registerSetting(setting);
```
Comment (reta) [inline server/src/main/java/org/opensearch/common/settings/SettingsModule.java:198]:
```suggestion
try {
            registerSetting(setting);
            if (onNodeSetting || onIndexSetting) {
                logger.info("Registered new Setting: " + setting.getKey() + " successfully ");
                return true;
            }
} catch (final IllegalArgumentException ex) {
            if (onNodeSetting) {
                clusterSettings.unregisterSetting(setting);
            }
            
            if (onIndexSetting) {
                indexScopedSettings.unregisterSetting(setting);
            }
}
```

Needs new `unregisterSetting` to compensate possible registration failures

Comment (ryanbogan) [reply server/src/main/java/org/opensearch/common/settings/SettingsModule.java:198]: I'm implementing unregisterSetting method in `AbstractScopedSettings.java` that removes the key from the respective settings maps.  Let me know if there is anything else that needs to be done in that method.

Comment (reta) [reply server/src/main/java/org/opensearch/common/settings/SettingsModule.java:198]: I think it should be fine, as far as settings are consistent, thank you

**Comment by reta** [comment]
@ryanbogan LGTM, I think we have only one comment left https://github.com/opensearch-project/OpenSearch/pull/5495#discussion_r1044701107

**Review by reta — APPROVED** [review]

**Review by saratvemulapalli — APPROVED** [review]

**Comment by ryanbogan** [comment]
Removed and added the backport label since it did not start the automatic backport process.

File: CHANGELOG.md [hunk 6]
```diff
@@ -16,6 +16,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 - Add feature flag for extensions ([#5211](https://github.com/opensearch-project/OpenSearch/pull/5211))
 - Added jackson dependency to server ([#5366] (https://github.com/opensearch-project/OpenSearch/pull/5366))
 - Added experimental extensions to main ([#5347](https://github.com/opensearch-project/OpenSearch/pull/5347))
+- Adding support to register settings dynamically ([#5495](https://github.com/opensearch-project/OpenSearch/pull/5495))
```
Comment (dblock) [inline CHANGELOG.md:19]: Should have been Add I think (we have all variations of added, adding and add :)

**Pull request merged**